### PR TITLE
Backup: copy-paste support with refactor of java side events handling

### PR DESCRIPTION
**Pull request by tpoisseau**
with bubbles

Refs: https://github.com/cheminfo/openchemlib-js/issues/229

### CODE REVIEW

File: lib/canvas_editor/events.js [hunk 1]
```diff
@@ -70,23 +70,27 @@ function addPointerListeners(canvasElement, drawArea, JavaEditorArea) {
 function addKeyboardListeners(canvasElement, editorArea, JavaEditorArea) {
   const isMac =
     typeof navigator !== 'undefined' && navigator.platform === 'MacIntel';
+  const isMenuKey = (ev) => (isMac && ev.metaKey) || (!isMac && ev.ctrlKey);
 
   function fireKeyEvent(what, ev) {
     const key = getKeyFromEvent(ev, JavaEditorArea);
     if (key === null) return;
-    ev.stopPropagation();
-    ev.preventDefault();
```
Comment (tpoisseau) [inline lib/canvas_editor/events.js:78]: Removed so browser can propagate copy-paste events.
To interact with clipboard, browser need proper user interaction

File: lib/canvas_editor/events.js [hunk 2]
```diff
@@ -70,23 +70,27 @@ function addPointerListeners(canvasElement, drawArea, JavaEditorArea) {
...
       ev.shiftKey,
-      (isMac && ev.metaKey) || (!isMac && ev.ctrlKey),
+      isMenuKey(ev),
     );
   }
 
   canvasElement.addEventListener('keydown', (ev) => {
+    // copy-paste is handled by the clipboard event
+    if (isMenuKey(ev) && ev.key === 'c') return;
+    if (isMenuKey(ev) && ev.key === 'v') return;
```
Comment (tpoisseau) [inline lib/canvas_editor/events.js:92]: not triggering java events with`fireKeyEvent` in this case, avoid double copy-paste events.
And events triggered from keyboard can't work. Need clipboard events

File: src/com/actelion/research/gwt/gui/generic/JSClipboardHandler.java [hunk 3]
```diff
@@ -19,72 +21,70 @@ private JavaScriptObject getJsHandler() {
...
+  // Can't be implemented here
+  // Clipboard is not accessible with sync API
+  public StereoMolecule pasteMolecule() {
+    return null;
+  }
+
+  // Not used, No implementation needed yet
   public StereoMolecule pasteMolecule(boolean prefer2D, int smartsMode) {
-    Object test = null;
-    System.out.println("test: " + test.toString());
```
Comment (tpoisseau) [inline src/com/actelion/research/gwt/gui/generic/JSClipboardHandler.java:35]: `test.toString()` produce an error (test is null).

Theses errors was really annoying to implement copy-paste properly.

So I just changes theses method so it's a no-op

**Comment by tpoisseau** [comment]
@targos I think it's working as expected, but I had to make lot of changes in Java side to implement copy-paste properly.

Do you agree on this approach ?
An other approach I can see is to do all the copy-paste system in JS side (and not propagate `ctrl+c`, `ctrl+v` events to java).

Actual approach is a bit hacky, (I also need to not propagate `ctrl+c`, `ctrl+v` to java). Doing in full JS will be hacky in the same way, and will be harder to replicate same java functionalities.

For me the best approach would be to refactor `IClipboardHandler` to support async operation (at least with callbacks), with an API which could be compatible with `ClipboardEvent` `DataTransfer` API.
I could manage the design of API and their implementations for JS side, but it will be hard for me to implement the java system (native) side.

Do you prefer I cut this PR into one for webcomponent and another one for copy-paste system ?

**Comment by targos** [comment]
Yes, please split the PR. We always prefer to do many smaller PRs if it makes sense.

**Comment by tpoisseau** [comment]
Please don't delete branch, backup for later works on copy-paste

**Pull request closed without merging**